src/main/java/org/truffleruby/core/FinalizationService.java

```diff
@@ -124,11 +130,34 @@ private void createFinalizationThread() {
             while (true) {
                 final FinalizerReference finalizerReference = (FinalizerReference) threadManager.runUntilResult(null,
                         () -> finalizerQueue.remove());
-                finalizerReference.getFinalizerActions().forEach(action -> action.run());
+                try {
+                    finalizerReference.getFinalizerActions().forEach(action -> action.run());
+                } catch (KillException e) {
+                    throw e;
+                } catch (RaiseException e) {
+                    handleFinalizeException(e);
+                } catch (Exception e) {
+                    // Do nothing, the finalizer thread must continue to process objects.
+                    if (context.getOptions().FINALIZE_PRINT_BACKTRACES) {
+                        e.printStackTrace();
+                    }
+                }
             }
         });
     }
 
+    private void handleFinalizeException(RaiseException exception) {
+        if (context.getOptions().FINALIZE_PRINT_BACKTRACES) {
```

Comment:
MRI shows
```
Exception `RuntimeError' at finalize.rb:10 - Oops
```
when an exception happens in a finalizer and `-d` is passed.
We should as well (see CoreExceptions#showExceptionIfDebug).